File: 028_ClutteredPaper.rb
# 028 - Cluttered Paper（★4）
# いもす法

M = 1_000
cnt = Array.new(M + 1) { [0] * (M + 1) }

N = gets.to_i
N.times do |i|
  lx, ly, rx, ry = gets.split.map(&:to_i)
  cnt[ly][lx] += 1
  cnt[ry][lx] -= 1
  cnt[ly][rx] -= 1
  cnt[ry][rx] += 1
end

2.times do |k|
  (k..M).each do |i|
    (1 - k..M).each do |j|
      cnt[i][j] += cnt[i - k][j - 1 + k]
    end
  end
end

ans = [0] * (N + 1)
cnt.flatten.each { ans[_1] += 1 }

puts ans[1..]

File: kotlin/028_ClutteredPaper.kt
/** 028 - Cluttered Paper（★4）
 * いもす法
 */

fun main() {
  val M = 1_000
  val N = readLine()!!.toInt()
  val cnt = MutableList(M + 1) { MutableList(M + 1) { 0 } }

  repeat(N) {
    val (lx, ly, rx, ry) = readLine()!!.split(" ").map(String::toInt)
    cnt[ly][lx] += 1
    cnt[ly][rx] -= 1
    cnt[ry][lx] -= 1
    cnt[ry][rx] += 1
  }

  for (i in 0 .. M) {
    for (j in 1 .. M) {
      cnt[i][j] += cnt[i][j - 1]
    }
  }

  for (i in 1 .. M) {
    for (j in 0 .. M) {
      cnt[i][j] += cnt[i - 1][j]
    }
  }

  val ans = MutableList(N + 1) { 0 }

  cnt.flatten().forEach { ans[it] += 1 }

  for (i in 1 .. N) println(ans[i])
}
